File: 0655-Print Binary Tree/0655-Print Binary Tree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
    vector<vector<string>> printTree(TreeNode* root) {
        int row = height(root);
        int col = (1 << row) - 1;
        vector<vector<string>> result(row, vector<string>(col));
        fill(result, root, 0, 0, col - 1);        
        return result;
    }
    
private:
    int height(TreeNode* root) {
        if (root == nullptr) {
            return 0;
        }
        
        return 1 + max(height(root->left), height(root->right));
    }
    
    void fill(vector<vector<string>>& result, TreeNode* root, int level, int l, int r) {
        if (root == nullptr) {
            return;
        }
        
        result[level][(l + r) / 2] = to_string(root->val);
        fill(result, root->left, level + 1, l, (l + r) / 2 - 1);
        fill(result, root->right, level + 1, (l + r) / 2 + 1, r);
    }
};

File: 0601-0700/0655-Print Binary Tree/0655-Print Binary Tree.py
# Definition for a binary tree node.
class TreeNode:
    def __init__(self, x):
        self.val = x
        self.left = None
        self.right = None

class Solution:
    def printTree(self, root: TreeNode) -> List[List[str]]:
        def height(root):
            if root is None:
                return 0
            return 1 + max(height(root.left), height(root.right))
        rows = height(root)
        cols = (1 << rows) - 1
        result = [[''] * cols for _ in range(rows)]
        def fill(node, level, l, r):
            if node is None:
                return
            mid = (l + r) // 2
            result[level][mid] = str(node.val)
            fill(node.left, level + 1, l, mid - 1)
            fill(node.right, level + 1, mid + 1, r)
        fill(root, 0, 0, cols - 1)
        return result
